File: fs/Fs.Problems/Problem004.fs
namespace Fs

    /// <title>Largest palindrome product</title>
    /// <summary>
    /// A palindromic number reads the same both ways. The largest palindrome
    /// made from the product of two 2-digit numbers is 9009 = 91 × 99.
    /// Find the largest palindrome made from the product of two 3-digit numbers.
    /// </summary>

    open System
    open System.Linq

    type Problem004 () =

        let IsPalindrome (s : string) =
            let half = s.Length / 2
            in Seq.forall2
                (fun x y -> x = y)
                (Seq.take half s) (Seq.take half (s.Reverse()))

        let NumberSolve =
            query {
                for i in Enumerable.Range(100, 899) do
                for j in Enumerable.Range(101, 899) do
                where (IsPalindrome (Convert.ToString (i*j)))
                select (i * j)
            } |> Seq.max

        member this.Solve () = NumberSolve |> string
        interface IProblem with
            member this.Solve () = this.Solve ()

File: euler/cs/Cs.Problems/001-009/Problem004.cs
using System.Linq;

namespace Cs.Problems
{
    /// <title>Largest palindrome product</title>
    /// <summary>
    /// A palindromic number reads the same both ways. The largest palindrome
    /// made from the product of two 2-digit numbers is 9009 = 91 × 99.
    /// Find the largest palindrome made from the product of two 3-digit numbers.
    /// </summary>
    /// <answer>906609</answer>
    public class Problem004 : IProblem
    {
        static bool IsPalindrome(string s)
        {
            for (int a = 0, b = s.Length - 1; a < b; a++, b--)
            {
                if (s[a] != s[b])
                {
                    return false;
                }
            }
            return true;
        }
        static int BruteSolve()
        {
            return (from i in Enumerable.Range(100, 899)
                    from j in Enumerable.Range(100, 899)
                    where IsPalindrome((i * j).ToString())
                    select i * j).Max();
        }

        public string Solve()
        {
            return BruteSolve().ToString();
        }
    }
}
